CMakeLists.txt

```diff
@@ -56,3 +56,31 @@ else()
 enable_testing()
 add_subdirectory(tests)
 endif()
+
+# first we can indicate the documentation build as an option and set it to OFF by default
+option(BUILD_DOC "Build documentation" OFF)
+
+# check if Doxygen is installed
+find_package(Doxygen)
+if (DOXYGEN_FOUND AND BUILD_DOC)
+    # set input and output files
+    set(DOXYGEN_IN ${CMAKE_CURRENT_SOURCE_DIR}/docs/Doxyfile.in)
+    set(DOXYGEN_OUT ${CMAKE_CURRENT_BINARY_DIR}/Doxyfile)
+    set(HaveDot NO)
+    if (DOXYGEN_DOT_FOUND)
+        set(HaveDot YES)
+    endif (DOXYGEN_DOT_FOUND)
+
+    # request to configure the file
+    configure_file(${DOXYGEN_IN} ${DOXYGEN_OUT} @ONLY)
+    message("Doxygen build started")
+
+    # note the option ALL which allows to build the docs together with the application
+    add_custom_target( doxygen ALL
```

Comment:
I would prefer this to be a `docs` target rather than `all` (following the principles in https://www.gnu.org/prep/standards/html_node/Standard-Targets.html).

Reply:
I have tried changing to the target to `html` but I cannot then figure out how to build the docs in VS 2019.

Reply:
I _think_ what we want is:

```cmake
add_custom_target(doxygen ...) # Without the ALL
add_custom_target(docs DEPENDS doxygen)
```

Reply:
This was convoluted to fix

Reply:
Thanks, I've added an ALL to the second line. First time I've done any high level CMake work in 6 months.